File: py/day01.py
import os
from collections import Counter

with open(f'../input/day01-input', 'r') as file:
    lines = [line.strip().split("  ") for line in file.readlines()]
    lines = [(int(parts[0]), int(parts[-1])) for parts in lines]

left1 = sorted(x[0] for x in lines)
right1 = sorted(x[1] for x in lines)

part1 = sum(abs(x - y) for x, y in zip(left1, right1))
print("Part 1:", part1)

left2 = Counter(left1)
right2 = Counter(right1)

part2 = sum(key * left2[key] * right2[key] for key in left2 if key in right2)
print("Part 2:", part2)

File: rs/src/days/day01.rs
use counter::Counter;
use crate::aocutils::*;

pub fn run() {
    let day = 1;
    let filename = find_input_file(&day).unwrap();

    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();

    match get_lines_from_file(&filename) {
        Ok(lines) => {
            for (index, line) in lines.iter().enumerate() {
                let numbers: Result<Vec<i32>, _> = line
                .split_whitespace()
                .map(|part| part.parse::<i32>())
                .collect();

                match numbers {
                    Ok(numbers) => {
                        left.push(numbers[0]);
                        right.push(numbers[1]);
                    }
                    Err(e) => eprintln!("Error parsing line {}: {}", index + 1, e),
                }
            }
        }
        Err(e) => eprintln!("Error reading file: {}", e),
    }

    left.sort();
    right.sort();
    let part1: i32 = left.iter().zip(right.iter()).map(|(x,y)| (x - y).abs() ).sum();
    print_day_result(&1,part1);

    let right2 = right.iter().collect::<Counter<_>>();
    let mut part2 = 0;
    left.iter().for_each(|x| {
        if right2.contains_key(x) {
            let rc = *right2.get(x).unwrap();
            part2 += x * rc as i32;
        }
    });
    print_day_result(&2, part2);
}
